src/main/java/cz/cvut/kbss/study/rest/PatientRecordController.java

```diff
@@ -83,6 +84,16 @@ public List<PatientRecordDto> getRecords(
         return result.getContent();
     }
 
+    @PreAuthorize("hasRole('" + SecurityConstants.ROLE_ADMIN + "') or @securityUtils.isMemberOfInstitution(#institutionKey)")
+    @GetMapping(value="availablePhases", produces = MediaType.APPLICATION_JSON_VALUE)
+    public Set<RecordPhase> getAvailableRecordPhases(@RequestParam(value = "institution", required = false) String institutionKey){
+        List<PatientRecord> records = recordService.findAll();
```

Comment:
Unfortunately, this will not work as it will basically load almost whole database. Each record contain the form which can be huge.

Basically `recordService.findAll()` is not used elsewhere for that reason, but instead `recordService.findAll(RecordFilterParams filters, Pageable pageSpec)` is used for that reason. You need to implement it using query, see `em.createNativeQuery` in source code.

Reply:
@blcham 
If I understand correctly, I should create a new method in PatientRecordDao (e.g. findAllAvailablePhases), which will execute the appropriate query to retrieve only distinct phases.

Reply:
yes